### ISSUE TITLE
Cannot read property 'minWordLength' of undefined, when using hyphenators

### ISSUE THREAD

**Issue by plagasul** [mnater/Hyphenopoly#92]
I am trying to use hyphenators to run hyphenopoly after ajax (fetch).

But I am consistently receiving a` Cannot read property 'minWordLength' of undefined` error. Both if firing the code from a script tag and also trying to run it on console:

![image](https://user-images.githubusercontent.com/13966673/75611582-0b004380-5b1c-11ea-80b1-55d528ecbfc7.png)

As you can see in the image, the element exists in the page, why is it then undefined ?

This is the initialization code :

```
<script>
  var Hyphenopoly = {
    require: {
	"it": "sovramagnificentissimamente",
	"en-gb": "Supercalifragilisticexpialidocious"
    },
    setup: {
	selectors: {
	    ".programme-title": {
		  hyphen: "•"
	    },
	}
    }
}
</script>
<script src="../assets/js/special/hyphenopoly/Hyphenopoly_Loader.js"></script>
```
 
And the hyphenator code

```
Hyphenopoly.hyphenators["en-gb"].then((hyphenator_en_gb) => {
    hyphenator_en_gb(document.querySelector("section.programme-text"));
});

```

Thank you

**Comment by mnater** [comment]
`hyphenator`-functions take an optional second argument denoting the selector (see: https://mnater.github.io/Hyphenopoly/Hyphenators.html) If that argument is not present it defaults to `.hyphenate` -- and you don't have settings for `.hyphenate`.

To solve the issue you have two possibilities:
a) define settings for a selector `.hyphenate`
or
b) tell your `hyphenator_en_gb`-function what selector to use for settings.

Hope this helps?
Kind regards,
Mathias

**Comment by plagasul** [comment]
Thank you for your quick answer!

Ah I see, I did not understand that from the documentation.

I still can't seem to make it work on AJAX, or perhaps I am not understanding the purpose of hyphenators.

This is the code atm:

```
<script>
  var Hyphenopoly = {
    require: {
	"it": "sovramagnificentissimamente",
	"en-gb": "Supercalifragilisticexpialidocious"
    },
    setup: {
	selectors: {
	    ".programme-title": {
	    },
	}
    }
}
</script>
<script src="../assets/js/special/hyphenopoly/Hyphenopoly_Loader.js"></script>
```

And the function to hyphenate on ajax load:

```
function hyphenate() {
  Hyphenopoly.hyphenators["en-gb"].then( function(hyphenator_en_gb) {
    hyphenator_en_gb(document.querySelectorAll(".programme-title"), ".programme-title");
  });
  Hyphenopoly.hyphenators["it"].then( function(hyphenator_it) {
    hyphenator_it(document.querySelectorAll(".programme-title"), ".programme-title");
  });
}
```

My expecation is that running `hyphenate()` after an ajax load will hyphenate  all `.programme-title` elements texts if it is needed.

I assumed that *not* providing a 'hyphen' option to the setup, selectors object would result in defaults being applied.

Thank you for your help.

**Comment by mnater** [comment]
Your code looks correct.

Keep in mind that Hyphenators are only available if the browser doesn't support the language or if you use FORCEHYPHENOPOLY. So you'd have to check if Hyphenopoly is active before calling your `hyphenate()`-function:

````javascript
if (Hyphenopoly) {
    hyphenate();
}
````

**Comment by plagasul** [comment]
Thank you.

I can't seem to be able to make it work. Not even with FORCEHYPHENOPOLY.

I must be missing some important bit of information. Let me describe how this page works.

All the initalization code is included in a html-pre.php snippet. I try to force Hyphenopoly and use a visible hyphen for debugging.

```
<script>
var Hyphenopoly = {
	require: {
		"it": "FORCEHYPHENOPOLY", 
		"en-gb": "FORCEHYPHENOPOLY"
	},
	setup: {
		selectors: {
			".programme-title": {
				hyphen: "•"
			},
		}
	}
	,
	paths: {
		// Use kirby to output here the right route
		patterndir: "<?= $kirby->url('assets') ?>/js/special/hyphenopoly/patterns/", //path to the directory of pattern files
		maindir: "<?= $kirby->url('assets') ?>/js/special/hyphenopoly/" //path to the directory where the other ressources are stored
	}
}
</script>
<?= js('assets/js/special/hyphenopoly/Hyphenopoly_Loader.js') ?> // path to the loader

```
The landing page does not contain any element with the selector '.programme-title'. After clicking anywhere in it, the home page is loaded which does contain '.programme-title' selectors, each labeled with a 'lang' attribut, which can be 'it' or 'en-gb', for example:

```
<li class="programme-title  " style="grid-area: 1 / 2" lang="en-gb">
  <div class="excerpt"><p>Vessel are Anna Santomauro , Viviana Vechia, Jaume Ferrete, Hanna Barbera....</p></div>
  <a href="http://localhost/~jaume/dev/vessel-art-project/it/vessel" data-ajax="" data-ajax-page="vessel" data-ajax-snippet="programme" data-ajax-loaded="" data-ajax-container=".flex-container" data-ajax-addmode="prepend" data-ajax-callback="" data-setclass="" data-setclass-target="body;body;body" data-setclass-class="programme;home;update" data-setclass-mode="add;remove;remove" data-urlstate="" data-urlstate-url="vessel" data-urlstate-title="Something canoe something firefly" data-title="Something canoe something firefly">Something canoe something firefly</a>
</li>
```

Right after the ajax content is loaded I call the 'hyphenate()' function:

```
function hyphenate() {
  Hyphenopoly.hyphenators["en-gb"].then( function(hyphenator_en_gb) {
    hyphenator_en_gb(document.querySelectorAll(".programme-title"), ".programme-title");
  });
  Hyphenopoly.hyphenators["it"].then( function(hyphenator_it) {
    hyphenator_it(document.querySelectorAll(".programme-title"), ".programme-title");
  });
}
```

But nothing seems to change in the page.

Some possibly noteworthy details:

- The state and selector of the en-gb or it promises before calling hyphenate()

![image](https://user-images.githubusercontent.com/13966673/76340970-ad7aac80-62fc-11ea-836a-174cde3276a2.png)

- The html lang attribute does not match Hyphenopoly locales, 'en_GB' instead of 'en-gb' and 'it_IT' instead of 'it' BUT the '.programme-title' elements lang attr do match hyphenopoly locales

- Even if I remove such lang attribute from the html element, and NO lang attributes exist on any element on the landing page I ALWAYS get this error, ONLY  for 'it'

`Hyphenopoly.js:211 engine for language 'it' loaded, but no elements found.`

- The '.programme-title' elements are cells of a CSS grid, which has this structure and CSS:

```
<ul class="grid"> // the grid container
  <li class="programme-index">...</li>
  <li class="programme-title">...</li> // This I want to hyphenate
  <li class="ticker">...</li>
  // Repeat a number of times
</ul>
```

```
.grid
  align-self: flex-start
  display: grid
  grid-template-columns: min-content 50% 50%
  grid-gap: 0px
  hyphens: auto
```

It looks like this, and as you can see the words don't break at all:

![image](https://user-images.githubusercontent.com/13966673/76342165-6beb0100-62fe-11ea-85a1-c396f9cf07bc.png)

Thank you very much for your help

**Comment by mnater** [comment]
Thanks for the detailed information. 
I have assembled a test case that does more or less what you describe:

````html
<!DOCTYPE html>
<html lang="en_GB">
    <head>
        <meta http-equiv="content-type" content="text/html; charset=UTF-8">
        <title>Test issue #92</title>
        <script>
            var Hyphenopoly = {
                require: {
                    "it": "FORCEHYPHENOPOLY", 
                    "en-gb": "FORCEHYPHENOPOLY"
                },
                setup: {
                    selectors: {
                        ".programme-title": {
                            hyphen: "•"
                        },
                    }
                }
            }
        </script>
        <script src="../Hyphenopoly_Loader.js"></script>
        <style type="text/css">
            .grid {
                align-self: flex-start;
                display: grid;
                grid-template-columns: min-content 50% 50%;
                grid-gap: 0px;
                hyphens: auto;
            }
        </style>
    </head>
    <body>
        <div id="inserthere"></div>
        <script>
            function hyphenate() {
                Hyphenopoly.hyphenators["en-gb"].then( function(hyphenator_en_gb) {
                    hyphenator_en_gb(document.querySelectorAll(".programme-title"), ".programme-title");
                });
                Hyphenopoly.hyphenators["it"].then( function(hyphenator_it) {
                    hyphenator_it(document.querySelectorAll(".programme-title"), ".programme-title");
                });
            }
            const htmlins = `
            <ul class="grid">
                <li class="programme-index">...</li>
                <li class="programme-title  " style="grid-area: 1 / 2" lang="en-gb">
                    <div class="excerpt"><p>Vessel are Anna Santomauro , Viviana Vechia, Jaume Ferrete, Hanna Barbera....</p></div>
                    <a href="http://localhost/~jaume/dev/vessel-art-project/it/vessel" data-ajax="" data-ajax-page="vessel" data-ajax-snippet="programme" data-ajax-loaded="" data-ajax-container=".flex-container" data-ajax-addmode="prepend" data-ajax-callback="" data-setclass="" data-setclass-target="body;body;body" data-setclass-class="programme;home;update" data-setclass-mode="add;remove;remove" data-urlstate="" data-urlstate-url="vessel" data-urlstate-title="Something canoe something firefly" data-title="Something canoe something firefly">Something canoe something firefly</a>
                </li>
                <li class="ticker">...</li>
            </ul>
            `;
            function insert() {
                document.getElementById("inserthere").innerHTML = htmlins;
                hyphenate();
            }
            window.setTimeout(insert(), 1000);
        </script>
    </body>
</html>
````

Unfortunately, I still can't reproduce the problem.

Is the website with the problem available somewhere? A link would be great to get to the bottom of the problem.

Does the problem occur in all browsers or is it specific?

**Comment by plagasul** [comment]
Thank you ,

Please visit https://www.vesselartproject.org/final/en/home from mobile or desktop in portrait mode.

It does seems to work on Safari ios and desktop, and on firefox desktop.

Does not seem to work on chrome android (unless caché is playing tricks on me) or chrome desktop (linux and mac).

**Comment by plagasul** [comment]
Actually, not exactly. What I am seeing in Safari ios and desktop, and on firefox desktop is hyphenation with proper hyphens, not with the hyphen defined in the selector above.

Additionally, if I inspect the element in Firefox, I don't see any soft hyphens in the code; I recall Hyphenopoly does introduced them visibly in the code.

![image](https://user-images.githubusercontent.com/13966673/76396915-6df2a580-637a-11ea-888d-937f67a36e00.png)

So, considering that I am forcing hyphenopoly should I assume it is not working at all anywhere?

[As this happened before without AJAX ](https://github.com/mnater/Hyphenopoly/issues/78), I've double checked the lang attributes I am using, and they seem valid IANA subtags.

Thank you

**Comment by mnater** [comment]
Hi

I think I found the problem. There are two bugs.

**Bug 1**
A `hyphenator` expects either a `String` or an `HTMLElement` as its first argument, but `document.querySelectorAll` returns a `NodeList`.

My testcase above has a bug (I called `insert()` instead of `insert` in the `setTimeout`, so the html got inserted immediately and got hyphenated directly and not by the hyphenator-function). Sorry about that stupid error...

So you'll have to unwrap the `nodeList` and call the hyphenator-function on each node:

````javascript
Hyphenopoly.hyphenators["en-gb"].then( function(hyphenator_en_gb) {
    const nl = document.querySelectorAll(".programme-title");
    nl.forEach((el) => {
        hyphenator_en_gb(el, ".programme-title");
    });
});
````

**Bug 2**
Hyphenators in their current implementation don't hyphenate childElements of the Element handed over to them.

I have either to document this or to implement recursive hyphenation of childElements in hyphenators somehow. Such an implementation is more complicated than it sounds - so it will take some time.

As a workaround you could hyphenate only elements with no childElements (aka the leaves of the DOM-Tree). Since there's afaik no selector to select elements with no children you'll have to add a class to those leaf-Elements and select this class.

Thank you for your patience and thanks again for the detailed report,
Kind regards,
Mathias

**Comment by plagasul** [comment]
Thank you again for your reply.

> **Bug 1**
> A `hyphenator` expects either a `String` or an `HTMLElement` as its first argument, but `document.querySelectorAll` returns a `NodeList`.

Ahá! I actually considered this at some point, but after reading the code had the impression it was able to handle node lists. I will do as you say here.

> **Bug 2**
> Hyphenators in their current implementation don't hyphenate childElements of the Element handed over to them.
> 
> As a workaround you could hyphenate only elements with no childElements (aka the leaves of the DOM-Tree). Since there's afaik no selector to select elements with no children you'll have to add a class to those leaf-Elements and select this class.

Can't I simply use a selector such as '.programme-title a' ?

Edit, I'll answer myself here: Yes, such a selector can be used, and it is enough for my use-case. But I can see how one may want to hyphenate several children at once, yes.

I believe this solves my problems,  thank you very much

I leave it open in any case, as you changed it to a bug

Thanks again

**Comment by mnater** [comment]
I'm glad I could help.

>Can't I simply use a selector such as '.programme-title a' ?
Edit, I'll answer myself here: Yes, such a selector can be used, and it is enough for my use-case. But I can see how one may want to hyphenate several children at once, yes.

I would prefer a clear separation between style and implementation. Therefore I have suggested a separate class.
I have created a new issue (https://github.com/mnater/Hyphenopoly/issues/96)

Regards,
Mathias
